File: 1.两数之和.js
/*
 * @lc app=leetcode.cn id=1 lang=javascript
 *
 * [1] 两数之和
 */
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// 遍历，o(n2)
// var twoSum = function(nums, target) {
//     for (let index = 0; index < nums.length; index++) {
//       const numA = nums[index];
//       for (let i = index + 1; i < nums.length; i++) {
//         if (numA + nums[i] === target) {
//           return [index, i];
//         }
//       }
//     }
// };

// 一次遍历，当前数的差不在已遍历的数中，则入表
// 使用哈希表map提升效率
var twoSum = function (nums, target) {
  const map = new Map();
  for (let i = 0; i < nums.length; i++) {
    let d = target - nums[i];
    if (map.has(d)) {
      return [map.get(d), i];
    }
    map.set(nums[i], i);
  }
  return [];
}

File: 1.两数之和.rs
/*
 * @lc app=leetcode.cn id=1 lang=rust
 *
 * [1] 两数之和
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut targets = HashMap::new();
        let mut result = Vec::new();
        for (i, v) in nums.iter().enumerate() {
            match targets.get(&(target - *v)) {
                Some(ii) => {
                    result.push(*ii as i32);
                    result.push(i as i32);
                    break;
                },
                _ => {
                    targets.insert(v, i);
                }
            }
        };
        return result;
    }
}
// @lc code=end

